test/controllers/users_controller_test.rb

```diff
@@ -32,4 +32,41 @@
       expect(flash[:result_text]).must_equal "User not found!"
     end
   end
+
+  describe "login" do
+    it "can log in an existing user" do
+      # Arrange
+      user_count = User.count
+
+      # Act
+      user = perform_login
+
+      expect(user_count).must_equal User.count
+
+      # Should also test Flash notices
+
+      expect(session[:user_id]).must_equal user.id
+    end
+
+    it "can log in a new user" do
+      # Arrange
+      user = User.new(provider: "github", username: "peter", uid: 987, email: "peter@griffin.com")
+
+      expect {
+        # Act
+        perform_login(user)
+        # Assert
+      }.must_change "User.count", 1
+
+      # Should also test Flash notices
+      user = User.find_by(uid: user.uid, provider: user.provider)
+
+      expect(session[:user_id]).must_equal user.id
+    end
+  end
+
```

Comment:
we should also test to make sure that if the user provides invalid info (like no name or whatever is required), we get the error messages ("Could not create new user account: #{user.errors.messages}") we are expecting and a redirect